src/Qir/Runtime/lib/QIR/QubitManager.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Qubit Manager maintains mapping between user qubit objects and
+// underlying qubit identifiers (Ids). When user program allocates
+// a qubit, Qubit Manager decides whether to allocate a fresh id or
+// reuse existing id that was previously freed. When user program
+// releases a qubit, Qubit Manager tracks it as a free qubit id.
+// Decision to reuse a qubit id is influenced by restricted reuse
+// areas. When a qubit id is freed in one section of a restricted
+// reuse area, it cannot be reused in other section of the same area.
+// Borrowing of qubits is not supported and is currently implemented
+// as plain allocation.
+
+#include <iostream>
+#include <vector>
+#include <limits>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
+
+    using QubitIdType = ::int32_t;
```

Comment:
Up to you: I would suggest to consider the _unsigned type_ for the unsigned concept - e.g. **u**int32_t.

Reply:
A note on the use of signed vs. unsigned integers for qubit ids.

First, I would like to establish that indexes and capacities should be of the same type. This is simply because they are compared. Comparing signed/unsigned will result in a mismatch.

Second, capacities and counts should be the same as well as they are also compared.

Third, the indexes and the values stored in the status array should be of the same type because I do store indexes directly.

So, indexes, stored values, counts and capacities should be of the same type. Can they be all unsigned? Yes. Can they all be signed? Yes. Both variants are possible. However, I would argue that signed type is more convenient.

One problem with unsigned types is that one of the typical iteration patterns doesn’t work. Specifically a loop that iterates indexes backwards may not always work:
`for (int i = count-1; i >= 0; i-- ) {}`
This will work perfectly for a signed type but will loop forever with an unsigned type.

Working with non-negative integers may occasionally result in a small negative integer (For example, x-1 < x is always true for non-negative signed integers). For signed integers, an absolute difference between any two positive integers will always fit into a signed integer (abs(index1-index2) is always what you want for non-negative signed integers). Counting items may occasionally get negative and a check count >= 0 is more intuitive that what it would have to be otherwise. All of this is because with signed integers you typically work in the middle of the range rather than on the boundary.

Another remark specific to the qubit manager is that the range of values stored in the array can be naturally split between non-negative and negative integers. Non-negative integers are used for free qubits and negative integers are used for other statuses. A simple comparison value >= 0 establishes if a qubit is free or not.

In the end, all of this is possible with unsigned integers and there’re different problems with signed integers that can be overlooked. But having worked with STL (which uses unsigned integers) and C# (which uses signed integers) I would prefer signed integers.

A few pointers to discussions can be found here: https://stackoverflow.com/questions/3658560/why-is-count-not-an-unsigned-integer

Reply:
Thank you very much, this is a very good topic to know.